File: modules/agent/src/controllers/notification_controller.py
"""Controller layer for notification routing to services."""

from fastapi import Request

from lib.decorators import handle_http_exceptions
from serializers.notification import notification_to_response
from services.notification_service import (
    MarkReadRequest,
    MarkEntityReadRequest,
    get_unread_count as get_unread_count_service,
    get_notifications as get_notifications_service,
    mark_as_read as mark_as_read_service,
    mark_entity_as_read as mark_entity_as_read_service,
)

# Re-export for routes
__all__ = ["MarkReadRequest", "MarkEntityReadRequest"]


@handle_http_exceptions
async def get_unread_count(request: Request) -> dict:
    """Get unread notification count for a user."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id
    count = await get_unread_count_service(user_id, tenant_id)
    return {"unread_count": count}


@handle_http_exceptions
async def get_notifications(request: Request, limit: int) -> dict:
    """Get notifications for a user."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id
    notifications = await get_notifications_service(user_id, tenant_id, limit)
    unread_count = await get_unread_count_service(user_id, tenant_id)
    return {
        "notifications": [notification_to_response(n) for n in notifications],
        "unread_count": unread_count,
    }


@handle_http_exceptions
async def mark_as_read(request: Request, data: MarkReadRequest) -> dict:
    """Mark specific notifications as read."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id
    updated = await mark_as_read_service(data.notification_ids, user_id, tenant_id)
    return {"status": "ok", "updated": updated}


@handle_http_exceptions
async def mark_entity_as_read(request: Request, data: MarkEntityReadRequest) -> dict:
    """Mark all notifications for an entity as read."""
    user_id = request.state.user_id
    tenant_id = request.state.tenant_id
    updated = await mark_entity_as_read_service(user_id, tenant_id, data.entity_type, data.entity_id)
    return {"status": "ok", "updated": updated}

File: modules/agent/internal/controllers/notification_controller.go
package controllers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"agent/internal/middleware"
	"agent/internal/schemas"
	"agent/internal/services"
)

// NotificationController handles notification HTTP requests
type NotificationController struct {
	notifService *services.NotificationService
}

// NewNotificationController creates a new notification controller
func NewNotificationController(notifService *services.NotificationService) *NotificationController {
	return &NotificationController{notifService: notifService}
}

// GetNotifications handles GET /notifications
func (c *NotificationController) GetNotifications(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	tenantID, _ := middleware.GetTenantID(r.Context())

	limit := parseLimit(r.URL.Query().Get("limit"), 20, 100)

	resp, err := c.notifService.GetNotifications(userID, tenantID, limit)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, resp)
}

// GetNotificationCount handles GET /notifications/count
func (c *NotificationController) GetNotificationCount(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	tenantID, _ := middleware.GetTenantID(r.Context())

	count, err := c.notifService.GetUnreadCount(userID, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]int64{"unread_count": count})
}

// MarkAsRead handles POST /notifications/mark-read
func (c *NotificationController) MarkAsRead(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	tenantID, _ := middleware.GetTenantID(r.Context())

	var req schemas.MarkReadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if len(req.NotificationIDs) == 0 {
		writeError(w, http.StatusBadRequest, "notification_ids required")
		return
	}

	updated, err := c.notifService.MarkAsRead(req.NotificationIDs, userID, tenantID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"status":  "ok",
		"updated": updated,
	})
}

// MarkEntityAsRead handles POST /notifications/mark-entity-read
func (c *NotificationController) MarkEntityAsRead(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		writeError(w, http.StatusUnauthorized, "unauthorized")
		return
	}
	tenantID, _ := middleware.GetTenantID(r.Context())

	var req schemas.MarkEntityReadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	if req.EntityType == "" || req.EntityID == 0 {
		writeError(w, http.StatusBadRequest, "entity_type and entity_id required")
		return
	}

	updated, err := c.notifService.MarkEntityAsRead(userID, tenantID, req.EntityType, req.EntityID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"status":  "ok",
		"updated": updated,
	})
}

func parseLimit(limitStr string, defaultVal, maxVal int) int {
	if limitStr == "" {
		return defaultVal
	}
	parsed, err := strconv.Atoi(limitStr)
	if err != nil {
		return defaultVal
	}
	if parsed <= 0 || parsed > maxVal {
		return defaultVal
	}
	return parsed
}
